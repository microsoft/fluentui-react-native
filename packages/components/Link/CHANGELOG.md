# Change Log - @fluentui-react-native/link

This log was last generated on Thu, 09 Apr 2020 18:39:15 GMT and should not be manually modified.

## 0.2.11
Thu, 09 Apr 2020 18:39:15 GMT

### Patches

- merge conflicts (ppatboyd@outlook.com)
## 0.2.9
Fri, 03 Apr 2020 20:40:51 GMT

### Patches

- switch the bin name from just-script to fluent-scripts to disambiguate names (jasonmo@microsoft.com)
## 0.2.7
Mon, 30 Mar 2020 16:27:00 GMT

### Patches

- Update main attribute to specify TypeScript index file. (kinhln@microsoft.com)
## 0.2.0
Mon, 16 Mar 2020 23:03:24 GMT

### Minor changes

- turn on publishing (jasonmo@microsoft.com)