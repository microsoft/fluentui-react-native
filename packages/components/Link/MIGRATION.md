# Link Migration

## Migration from v0 Link (Once moved from Experimental to Components)

In the short term, the new `Link` control is named `LinkV1` while it clashes with the existing older control. Once we deprecate the old control, it will be renamed to `Link`. It may be useful to rename the control to `Link` using the import syntax to simplify the rename:

```ts
import { LinkV1 as Link } from '@fluentui-react-native/text';
```

### Component renames

| v0 `Link`                                                           | v1 `Link`                                                            |
| ------------------------------------------------------------------- | -------------------------------------------------------------------- |
| `<Link url="https://www.bing.com/" content="Click to navigate." />` | `<Link inline url="https://www.bing.com/">Click to Navigate.</Link>` |

### Props unchanged

- Any props that are part of `ViewProps` and `TextProps` from `react-native`
- `url`
- `componentRef`

### Props added

- `appearance`
- `enableFocusRing`
- `inline`
- `tooltip`

### Props removed

- `content` is not supported in the v1 Link. Text to display as Link is passed in as a child of the component

### Props only supported natively on win32 and windows

- `onKeyUp`
- `onKeyDown`
- `keyUpEvents`
- `keyDownEvents`
- `onMouseEnter`
- `onMouseLeave`

### Tokens unchanged

- Any props that are part of `FontTokens` as `TextTokens` and `IForegroundColorTokens`

### Tokens changed

- Event states are added as `LinkTokens`

| Token      | Win32      |
| ---------- | ---------- |
| `hovered`  | LinkTokens |
| `pressed`  | LinkTokens |
| `focused`  | LinkTokens |
| `visited`  | LinkTokens |
| `disabled` | LinkTokens |
| `inline`   | LinkTokens |
| `subtle`   | LinkTokens |

### Updating ThemeProvider

If you are using the older theme provider `ThemeProvider` from `@uifabricshared/theming-react-native`, you will need to update the `ThemeProvider` to pull from `@fluentui-react-native/theme` to have the control work properly with themes. Please see [this page](../../../docs/pages/Guides/UpdateThemeProvider.md) for guidance.

### Migrating customized Link

Please see [this page](../../../docs/pages/Guides/UpdatingCustomize.md) for guidance on how to move from the old `customize` API to the new one.

### Slot differences

There are no differences between block and inline Links
Note: Inline Links are not supported on win32.

| Default Link |             |
| ------------ | ----------- |
| Slot         | Type        |
| -----------  | ----------- |
| root         | `View`      |
| content      | `Text`      |

| Inline Link |             |
| ----------- | ----------- |
| Slot        | Type        |
| ----------- | ----------- |
| content     | `Text`      |
