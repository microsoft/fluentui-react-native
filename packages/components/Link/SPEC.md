# Link

(Once moved from Experimental to Components)
In the short term, the new `Link` control is named `LinkV1` while it clashes with the existing older control. Once we deprecate the old control, it will be renamed to `Link`. It may be useful to rename the control to `Link` using the import syntax to simplify the rename:

```ts
import { LinkV1 as Link } from '@fluentui-react-native/link';
```

FURN components that have imported from experimental-link will use the above syntax to use `LinkV1`.

## Background

The `Link` component is a clickable control primarily used for navigation, providing an interactive reference to a resource. It is usually displayed as an block element by default, but can be displayed inline as well.

## Requirements

If using FURN's theming, the `Link` requires use of the `ThemeProvider` from `@fluentui-react-native/theme` to work properly with themes. Please see [this page](../../../docs/pages/Guides/UpdateThemeProvider.md) for information on updating your `ThemeProvider` if using the version from `@uifabricshared/theming-react-native`.

## Sample Code

Basic example:

```jsx
<Link url="https://www.bing.com/">Click to Navigate.</Link>
```

More examples on the [Test pages for Link](../../../apps/fluent-tester/src/TestComponents/LinkV1). Instructions on running the tester app can be found [here](../../../apps/fluent-tester/README.md).

## Visual Examples

Win32:

![Default Links on win32 example](./assets/Link_example_default_win32.png)

```jsx
<Link tooltip={'https://www.bing.com/'} url="https://www.bing.com/">
  Click to navigate.
</Link>
<Link onPress={doPress} onAccessibilityTap={doAllyTap}>
  Click to alert.
</Link>
<Link disabled focusable>
  Disabled focusable Link
</Link>
```

![Subtle Links on win32 example](./assets/Link_example_subtle_win32.png)

```jsx
<Link appearance="subtle" url="https://www.bing.com/">
  Click to navigate.
</Link>
<Link appearance="subtle" onPress={doPress} disabled>
  Disabled Link
</Link>
```

![Customized Links on win32 example](./assets/Link_example_customized_win32.png)

```jsx
const BlueHeaderBold = Link.customize({ variant: 'headerStandard', fontWeight: '700', color: '#0229c4', textAlign: 'right' });
<BlueHeaderBold>BlueHeaderBold</BlueHeaderBold>;
```

## API

The default `Link` control is rendered as a FURN Text(`Text`) wrapped in a `react-native` View. However, inline `Link` is rendered purely as a FURN Text(`Text`) without a wrapping View.

### Props

Below is the set of props the Text supports:

```ts
export interface LinkProps extends IWithPressableOptions<TextProps> {
  /**
   * The appearance of the link, either `default` or `subtle`
   * Note: 'subtle' is not supported for Android
   * @default default
   */
  appearance?: LinkAppearance;
  /**
   * A RefObject to access the IButton interface. Use this to access the public methods and properties of the component.
   */
  componentRef?: React.RefObject<IFocusable>;
  /**
   * Whether to use native focus visuals for the component
   * @default true
   */
  enableFocusRing?: boolean;
  /**
   * Whether the link is inline with text
   * Note: Not supported for win32
   * @default false
   */
  inline?: boolean;
  /**
   * The URL that is opened when the link is clicked.  This value supersedes the 'onPress' callback when both are present.
   * @default undefined
   */
  url?: string;
  /**
   * Text that should show in a tooltip when the user hovers over a button.
   * Note: Not supported for Android
   */
  tooltip?: string;
}
```

### Styling Tokens

Tokens can be used to customize the styling of the control by using the `customize` function on the `Link`. For more information on using the `customize` API, please see [this page](../../framework/composition/README.md). The `Link` has the following tokens:

```ts
export interface LinkTokens extends TextTokens {
  hovered?: LinkTokens;
  pressed?: LinkTokens;
  focused?: LinkTokens;
  visited?: LinkTokens;
  disabled?: LinkTokens;
  inline?: LinkTokens;
  subtle?: LinkTokens;
}
```

Note: 'hovered','focused','visited','subtle' are not supported for Android

## Behaviors

### Link with url

`Text` can be rendered as pressable.

### Interaction

#### Keyboard interaction

The following is a set of keys that interact with the `Link` component:

| Key     | Description                                                                                         |
| ------- | --------------------------------------------------------------------------------------------------- |
| `Tab`   | If `disabled` is not set, component will gain focus.                                                |
| `Space` | Nagivates to the path passed in as a `url` or executes the function passed into the `onPress` prop. |

#### Cursor interaction for Link

- Cursor moves onto Link: Should immediately change the styling of the `Link` so that it appears to be hovered. Cursor changes to `pointer`.
- Cursor moves out of Link: Should immediately remove the hovered styling of the `Link`. Cursor changes to `arrow`.
- Cursor moves onto disabled Link: Cursor changes to `not-allowed`.
- Mouse click: Navigate to `url` or should handle `onPress` event of `Link` and move focus to its target.

#### Touch interaction

The same behavior as above translated for touch events. This means that there is no equivalent for `onHoverIn` and `onHoverOut`, which makes it so that the hovered state cannot be accessed.

## Accessibility

### Expected behavior

- accessibleRole: `link`
- Should mix in the accessibility props expected for a `Link` component.
- Should be keyboard tabbable and focusable.
- onAccessibilityTap defaults to onPress unless set otherwise
