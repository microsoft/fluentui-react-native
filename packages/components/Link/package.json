{
  "name": "@fluentui-react-native/link",
  "version": "0.8.19",
  "description": "A cross-platform Link component using the Fluent Design System",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "typings": "lib/index.d.ts",
  "onPublish": {
    "main": "lib-commonjs/index.js",
    "module": "lib/index.js"
  },
  "scripts": {
    "build": "fluentui-scripts build",
    "just": "fluentui-scripts",
    "clean": "fluentui-scripts clean",
    "depcheck": "fluentui-scripts depcheck",
    "lint": "fluentui-scripts eslint",
    "test": "fluentui-scripts jest",
    "update-snapshots": "fluentui-scripts jest -u",
    "verify-api": "fluentui-scripts verify-api-extractor",
    "update-api": "fluentui-scripts update-api-extractor"
  },
  "dependencies": {
    "@uifabricshared/foundation-compose": "^1.6.88",
    "@fluentui-react-native/adapters": ">=0.3.36 <1.0.0",
    "@fluentui-react-native/interactive-hooks": ">=0.5.2 <1.0.0",
    "@fluentui-react-native/text": ">=0.6.16 <1.0.0",
    "@fluentui-react-native/tokens": ">=0.5.13 <1.0.0",
    "@uifabricshared/foundation-composable": ">=0.6.75 <1.0.0",
    "@uifabricshared/foundation-settings": ">=0.7.2 <1.0.0"
  },
  "devDependencies": {
    "@types/react-native": "^0.62.0",
    "@uifabricshared/build-native": "^0.1.1",
    "@uifabricshared/eslint-config-rules": "^0.1.1",
    "react-native": "0.62.2"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-native": ">=0.60.0"
  },
  "author": "",
  "license": "MIT"
}
