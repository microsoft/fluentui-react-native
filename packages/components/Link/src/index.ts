export { linkName as linkNameV1 } from './Link.types';
export type { LinkTokens, LinkProps, LinkAppearance, LinkState, LinkSlotProps, LinkType } from './Link.types';
export { defaultLinkTokens } from './LinkTokens';
export { linkStates, stylingSettings as linkStylingSettings } from './Link.styling';
export { useLink } from './useLink';
export { Link as LinkV1, linkLookup } from './Link';

export { linkName } from './legacy/Link.types';
export type {
  ILinkInfo,
  ILinkOptions,
  ILinkProps,
  ILinkRenderData,
  ILinkSlotProps,
  ILinkState,
  ILinkTokens,
  ILinkType,
  IWithLinkOptions,
} from './legacy/Link.types';
export { Link, useAsLink } from './legacy/Link';
export type { ILinkHooks } from './legacy/Link';
