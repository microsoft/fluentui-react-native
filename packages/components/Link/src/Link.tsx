/** @jsx withSlots */
import * as React from 'react';

import { Linking, View } from 'react-native';
import { Text } from '@fluentui-react-native/text';
import { compose, IUseComposeStyling } from '@uifabricshared/foundation-compose';
import { ILinkProps, ILinkSlotProps, ILinkState, ILinkRenderData, IWithLinkOptions, linkName, ILinkType } from './Link.types';
import { settings } from './Link.settings';
import { foregroundColorTokens, textTokens, borderTokens } from '@fluentui-react-native/tokens';
import { useAsPressable, useKeyCallback, useViewCommandFocus } from '@fluentui-react-native/interactive-hooks';
import { mergeSettings } from '@uifabricshared/foundation-settings';
import { ISlots, withSlots } from '@uifabricshared/foundation-composable';
import { IViewProps } from '@fluentui-react-native/adapters';

export type ILinkHooks = [IWithLinkOptions<IViewProps>, ILinkState];

export function useAsLink(userProps: IWithLinkOptions<IViewProps>): ILinkHooks {
  const { url, onPress, ...rest } = userProps;

  const [linkState, setLinkState] = React.useState({ visited: false });
  const linkOnPress = React.useCallback(
    (e) => {
      setLinkState({ visited: true });
      if (url) {
        Linking.openURL(url as string);
      } else if (onPress) {
        onPress(e);
      }
    },
    [setLinkState, url, onPress],
  );
  const pressable = useAsPressable({ onPress: linkOnPress, ...rest });
  const onKeyUp = useKeyCallback(linkOnPress, ' ', 'Enter');

  const newState = {
    ...pressable.state,
    ...linkState,
  };

  const newProps = {
    ...userProps,
    ...pressable.props,
    onKeyUp,
  };

  return [newProps, newState];
}

export const Link = compose<ILinkType>({
  displayName: linkName,
  settings,
  usePrepareProps: (userProps: ILinkProps, useStyling: IUseComposeStyling<ILinkType>): ILinkRenderData => {
    const { content, onAccessibilityTap, ...rest } = userProps;

    const [linkProps, linkState] = useAsLink(rest);
    const onAccTap = onAccessibilityTap ? onAccessibilityTap : linkProps.onPress;

    const info = { content: !!content };

    const linkRef = useViewCommandFocus(userProps.componentRef);

    // grab the styling information, referencing the state as well as the props
    const styleProps = useStyling(userProps, (override: string) => linkState[override] || userProps[override]);

    // create the merged slot props
    const slotProps = mergeSettings<ILinkSlotProps>(styleProps, {
      root: { ...linkProps, ref: linkRef, onAccessibilityTap: onAccTap },
      content: { children: content },
    });

    return { slotProps, state: { ...linkState, ...info } };
  },
  render: (Slots: ISlots<ILinkSlotProps>, renderData: ILinkRenderData, ...children: React.ReactNode[]) => {
    const content = renderData.state && renderData.state.content;

    return children && children.length && children.length === 1 && children[0] !== undefined ? (
      <Slots.root>
        {content && <Slots.content />}
        {children}
      </Slots.root>
    ) : (
      <Slots.root>{content && <Slots.content />}</Slots.root>
    );
  },
  slots: {
    root: View,
    content: Text,
  },
  styles: {
    root: [],
    content: [foregroundColorTokens, textTokens, borderTokens],
  },
});

export default Link;
