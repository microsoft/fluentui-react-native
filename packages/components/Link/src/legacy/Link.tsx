/** @jsxRuntime classic */
/** @jsx withSlots */
import * as React from 'react';
import { Linking, View } from 'react-native';

import type { IViewProps } from '@fluentui-react-native/adapters';
import { useAsPressable, useKeyProps, useOnPressWithFocus, useViewCommandFocus } from '@fluentui-react-native/interactive-hooks';
import { Text } from '@fluentui-react-native/text';
import { foregroundColorTokens, textTokens, borderTokens } from '@fluentui-react-native/tokens';
import type { ISlots } from '@uifabricshared/foundation-composable';
import { withSlots } from '@uifabricshared/foundation-composable';
import type { IUseComposeStyling } from '@uifabricshared/foundation-compose';
import { compose } from '@uifabricshared/foundation-compose';
import { mergeSettings } from '@uifabricshared/foundation-settings';

import { settings } from './Link.settings';
import type { ILinkProps, ILinkSlotProps, ILinkState, ILinkRenderData, IWithLinkOptions, ILinkType } from './Link.types';
import { linkName } from './Link.types';

export type ILinkHooks = [IWithLinkOptions<IViewProps>, ILinkState];

export function useAsLink(userProps: IWithLinkOptions<IViewProps>, ref: React.RefObject<any>): ILinkHooks {
  const { url, onPress, ...rest } = userProps;

  const [linkState, setLinkState] = React.useState({ visited: false });
  const linkOnPress = React.useCallback(
    (e) => {
      setLinkState({ visited: true });
      if (url) {
        Linking.openURL(url as string);
      } else if (onPress) {
        onPress(e);
      }
    },
    [setLinkState, url, onPress],
  );

  // Ensure focus is placed on link after click
  const linkOnPressWithFocus = useOnPressWithFocus(ref, linkOnPress);

  const pressable = useAsPressable({ onPress: linkOnPressWithFocus, ...rest });
  const onKeyUpProps = useKeyProps(linkOnPress, ' ', 'Enter');

  const newState = {
    ...pressable.state,
    ...linkState,
  };

  const newProps = {
    ...userProps,
    ...pressable.props,
    ...onKeyUpProps,
  };

  return [newProps, newState];
}

export const Link = compose<ILinkType>({
  displayName: linkName,
  settings,
  usePrepareProps: (userProps: ILinkProps, useStyling: IUseComposeStyling<ILinkType>): ILinkRenderData => {
    const defaultComponentRef = React.useRef(null);
    const { content, onAccessibilityTap, componentRef = defaultComponentRef, ...rest } = userProps;

    const [linkProps, linkState] = useAsLink(rest, componentRef);
    const onAccTap = onAccessibilityTap ? onAccessibilityTap : linkProps.onPress;

    const info = { content: !!content };

    const linkRef = useViewCommandFocus(componentRef);

    // grab the styling information, referencing the state as well as the props
    const styleProps = useStyling(userProps, (override: string) => linkState[override] || userProps[override]);

    // create the merged slot props
    const slotProps = mergeSettings<ILinkSlotProps>(styleProps, {
      root: { ...linkProps, ref: linkRef, onAccessibilityTap: onAccTap },
      content: { children: content },
    });

    return { slotProps, state: { ...linkState, ...info } };
  },
  render: (Slots: ISlots<ILinkSlotProps>, renderData: ILinkRenderData, ...children: React.ReactNode[]) => {
    const content = renderData.state && renderData.state.content;

    return children && children.length && children.length === 1 && children[0] !== undefined ? (
      <Slots.root>
        {content && <Slots.content />}
        {children}
      </Slots.root>
    ) : (
      <Slots.root>{content && <Slots.content />}</Slots.root>
    );
  },
  slots: {
    root: View,
    content: Text,
  },
  styles: {
    root: [],
    content: [foregroundColorTokens, textTokens, borderTokens],
  },
});

export default Link;
