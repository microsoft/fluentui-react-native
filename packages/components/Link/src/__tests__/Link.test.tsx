import * as React from 'react';
import { Alert } from 'react-native';

import { checkRenderConsistency, checkReRender } from '@fluentui-react-native/test-tools';
import * as renderer from 'react-test-renderer';

import { Link } from '../Link';

describe('Link component tests', () => {
  it('Default Link', () => {
    const tree = renderer.create(<Link url="https://www.bing.com">Link to Bing</Link>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Default Link as a pressable', () => {
    const doPress = (): void => {
      Alert.alert('Alert.', 'You have been alerted.');
    };
    const tree = renderer.create(<Link onPress={doPress}>Link to Bing</Link>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Inline Link', () => {
    const tree = renderer
      .create(
        <Link inline url="https://www.bing.com">
          Link to Bing
        </Link>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Subtle Link', () => {
    const tree = renderer
      .create(
        <Link appearance="subtle" url="https://www.bing.com">
          Link to Bing
        </Link>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Link simple rendering does not invalidate styling', () => {
    checkRenderConsistency(() => <Link url="https://www.bing.com">Link to Bing</Link>, 2);
  });

  it('Link re-renders correctly', () => {
    checkReRender(() => <Link url="https://www.bing.com">Render twice</Link>, 2);
  });

  it('Link shares produced styles across multiple renders', () => {
    const style = { color: 'black' };
    checkRenderConsistency(
      () => (
        <Link style={style} url="https://www.bing.com">
          Shared styles
        </Link>
      ),
      2,
    );
  });

  it('Link re-renders correctly with style', () => {
    const style = { color: 'black' };
    checkReRender(
      () => (
        <Link style={style} url="https://www.bing.com">
          Shared Style Render
        </Link>
      ),
      2,
    );
  });
});
