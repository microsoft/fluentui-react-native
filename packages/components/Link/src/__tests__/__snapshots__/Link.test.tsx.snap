// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component tests Default Link 1`] = `
<View
  accessibilityRole="link"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  cursor="pointer"
  disabled={false}
  enableFocusRing={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "flex-start",
    }
  }
  tooltip="https://www.bing.com"
>
  <Text
    ellipsizeMode="tail"
    focusable={false}
    numberOfLines={0}
    style={
      {
        "alignSelf": "flex-start",
        "color": "#0f6cbd",
        "fontFamily": "Segoe UI",
        "fontSize": 13,
        "fontWeight": "400",
        "margin": 0,
      }
    }
  >
    Link to Bing
  </Text>
</View>
`;

exports[`Link component tests Default Link as a pressable 1`] = `
<View
  accessibilityRole="link"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  cursor="pointer"
  disabled={false}
  enableFocusRing={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "flex-start",
    }
  }
>
  <Text
    ellipsizeMode="tail"
    focusable={false}
    numberOfLines={0}
    style={
      {
        "alignSelf": "flex-start",
        "color": "#0f6cbd",
        "fontFamily": "Segoe UI",
        "fontSize": 13,
        "fontWeight": "400",
        "margin": 0,
      }
    }
  >
    Link to Bing
  </Text>
</View>
`;

exports[`Link component tests Inline Link 1`] = `
<View
  accessibilityRole="link"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  cursor="pointer"
  disabled={false}
  enableFocusRing={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "flex-start",
    }
  }
  tooltip="https://www.bing.com"
>
  <Text
    ellipsizeMode="tail"
    focusable={false}
    numberOfLines={0}
    style={
      {
        "alignSelf": "flex-start",
        "color": "#0f6cbd",
        "fontFamily": "Segoe UI",
        "fontSize": 13,
        "fontWeight": "400",
        "margin": 0,
        "textDecorationLine": "underline",
      }
    }
  >
    Link to Bing
  </Text>
</View>
`;

exports[`Link component tests Subtle Link 1`] = `
<View
  accessibilityRole="link"
  accessibilityState={
    {
      "disabled": false,
    }
  }
  accessible={true}
  appearance="subtle"
  cursor="pointer"
  disabled={false}
  enableFocusRing={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onPress={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignSelf": "flex-start",
    }
  }
  tooltip="https://www.bing.com"
>
  <Text
    ellipsizeMode="tail"
    focusable={false}
    numberOfLines={0}
    style={
      {
        "alignSelf": "flex-start",
        "color": "#0f6cbd",
        "fontFamily": "Segoe UI",
        "fontSize": 13,
        "fontWeight": "400",
        "margin": 0,
      }
    }
  >
    Link to Bing
  </Text>
</View>
`;
